Contour lines

romanshuvalov opened the pull request:
Added `import-contour`. Note that contour lines MUST be imported, otherwise export script may fail because of absence of contour-related tables.

lukasmartinelli commented:
Great @romanshuvalov. Really appreciate it. Most important thing is whether the tile size is still acceptable that we can merge it into the main vector tiles.

- [ ] I will test this out with Switzerland and generate MBTiles and then compare.

After that is ok we can improve the PR or otherwise move it to the dem2vectorterrain project and create vector tiles just for the contour lines (copy the structure from https://github.com/osm2vectortiles/custom-vt-skeleton).

> Added import-contour. Note that contour lines MUST be imported, otherwise export script may fail because of absence of contour-related tables.

We can fix that by just creating a empty table before hand. Then people without shape files just don't have contours.

lukasmartinelli commented:
Okay size impact seems okay on the individual tile level (of course the planet will be bigger).
But it won't do damage to the 500KB tile limit (additional 2-5KB per z14 tile).

Speed impact is bigger though -rendering is quite a bit slower.
Next steps would be to test this on a bigger scale than Switzerland (and also check how big the PostGIS db get's with all those shapefiles).

Hmm thinking whether it belongs to OSM2VectorTiles or not. Your changes will be available as vector tiles for sure @romanshuvalov. It is so great!! But question is whether a separate vector tile set or not.

This is a vector tile set that actually only needs to be generated one time only (no update functionality as for OSM2VectorTiles). 
But on the other hand integrating it into OSM2VectorTiles **makes it immediately visible for the people and they will actually start using it**.

@klokan let's get together at FOSS4G and discuss it.
I hope you don't mind @romanshuvalov if this takes some time to discuss.
And also when we are going to render a global vector tile set this will also take some time but it is a huge benefit to everyone!

hyperknot commented:
If you decide to keep them in a separate set (which might not need to be updated for a long long time), I'd recommend merging them into published mbtiles databases, so clients would not need to use 2x the requests.

lukasmartinelli commented:
> If you decide to keep them in a separate set (which might not need to be updated for a long long time), I'd recommend merging them into published mbtiles databases, so clients would not need to use 2x requests for each tile.
>  Hide all checks

@hyperknot you can also concat PBFs on the client side. Petr's https://github.com/klokantech/tileserver-gl does this (as does Mapbox).
However not every tileserver can do that and als makes hosting from S3 more difficult.

What are your thoughts on this @hyperknot ?

hyperknot commented:
I was referring clients as in browsers and mobile devices. When mobile devices need to make 2x (or 3x for hillshading) requests, that gives a visible latency especially on mobile networks. HTTP2 can help a bit, but it't not yet common.

I've just checked Mapbox outdoors, and they only serve a single 512x512 PBF which has both vector data, contour lines, and triangulated hillshading all combined. Thus even a desktop browser only does 6 requests for a single map view, which is quite fast.

If this can be done on-the-fly via TileServer GL without any performance penalties, then I guess it's a good solution. If this means much higher system requirements / slower serving performance, then I believe an offline processing step would be better.

romanshuvalov commented:
Best solution is render two types of vector tiles: with contours and without them. So probably we should add options to  `export`: render only osm data, render osm data + contour layer, render both. 

Contours as separate tile set is not good because of 2x HTTP queries and in most cases they don't need without OSM data tiles. But it require more drive space because of doubling OSM data. It can be solved by adding third option **contours-only**. So it could be configured by administrator.

Which tile set - pure OSM or OSM+contours - should be given on osm2vectortiles website, it's for you to decide. I think OSM+contours are better. 

About me: for my purposes (3D rendering) I need custom tiles with a lot of other fields and calculations so I forked the project and I will continue developing in separated branch. So, of course, I don't mind if you need time to deal with that questions. Do whatever you want, I just needed contour layer and I added it and shared the code in order to be useful for community. 

If you decide something and you think I could help implementing it, I will participate. Meanwhile I continue working on my own project.

romanshuvalov commented:
**Update**: Fixed Makefile. And added empty tables in case of no contour files in `import/contour` folder.

So, contour files are not necessary now, but `import-contour` still must be called for creating that empty tables.

And I don't know how to create empty `import/contour/` folder on github. It's listed in `.gitignore` file and I can't add it to commit.

stirringhalo commented:
@lukasmartinelli @romanshuvalov 

My personal preference is to have both Contours alone and OSM+Contours. With https://github.com/stirringhalo/openstreetmap-mapboxgl coming along well (waiting on property functions for line features so I can render all roads in one go), I'd like to merge in the contours there too.
